File: tasks/03-sync-aggregator/python/aggregator.py
"""
Concurrent File Stats Processor – Python stub.

Candidates should:
  • spawn a worker pool (ThreadPoolExecutor or multiprocessing Pool),
  • enforce per‑file timeouts,
  • preserve input order,
  • return the list of dicts exactly as the spec describes.
"""
from __future__ import annotations
from typing import List, Dict


def aggregate(filelist_path: str, workers: int = 4, timeout: int = 2) -> List[Dict]:
    """
    Process every path listed in *filelist_path* concurrently.

    Returns a list of dictionaries in the *same order* as the incoming paths.

    Each dictionary must contain:
        {"path": str, "lines": int, "words": int, "status": "ok"}
    or, on timeout:
        {"path": str, "status": "timeout"}

    Parameters
    ----------
    filelist_path : str
        Path to text file containing one relative file path per line.
    workers : int
        Maximum number of concurrent worker threads.
    timeout : int
        Per‑file timeout budget in **seconds**.
    """
    # ── TODO: IMPLEMENT ──────────────────────────────────────────────────────────
    raise NotImplementedError("implement aggregate()")
    # ─────────────────────────────────────────────────────────────────────────────

File: tasks/03-sync-aggregator/csharp/Aggregator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

public static class Aggregator
{
    public record Result(string Path, int Lines, int Words, string Status);

    /// <summary>
    /// Concurrently processes the files listed in <paramref name="fileListPath"/>.
    /// Must preserve input order and apply a per‑file timeout.
    /// </summary>
    public static IEnumerable<Result> Aggregate(
        string fileListPath,
        int workers = 4,
        int timeout = 2)
    {
        // ── TODO: IMPLEMENT ────────────────────────────────────────────────────
        throw new NotImplementedException("implement Aggregate()");
        // ───────────────────────────────────────────────────────────────────────
    }
}
